### PR TITLE
Add karmadactl e2e test

### PR DESCRIPTION
**Pull request by wuyingjun-lucky**
Signed-off-by: wuyingjun <wuyingjun_yewu@cmss.chinamobile.com>

**What type of PR is this?**
/kind failing-test

**What this PR does / why we need it**:
Add karmadactl e2e test 
**Which issue(s) this PR fixes**:
Fixes #https://github.com/karmada-io/karmada/issues/1786

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

File: test/e2e/karmadactl_test.go [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+
+		ginkgo.AfterEach(func() {
+			framework.RemoveDeployment(kubeClient, deploymentNamespace, deploymentName)
+			framework.RemovePropagationPolicy(karmadaClient, deploymentNamespace, deploymentName+"-propagation")
+			framework.RemoveNamespace(kubeClient, deploymentNamespace)
+		})
+
+		ginkgo.It("Test promoting a deployment from cluster member", func() {
+
+			//  Step 1,  create ns and deploy on cluster member1
```
Comment (XiShanYongYe-Chang) [inline test/e2e/karmadactl_test.go:62]: We usually add a space at the beginning of the comment, and the same goes for other places.

File: test/e2e/karmadactl_test.go [hunk 2]
```diff
@@ -0,0 +1,330 @@
...
+					ClusterNamespace: options.DefaultKarmadaClusterNamespace,
+				}
+				args := []string{"deployment", deploymentName}
+				//init args: place deployment name to CommandPromoteOption.name
+				err := opts.Complete(args)
+				gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+
+				//use karmadactl to promote a deployment from member1 and it will failed because the ns is not existed
+				err = karmadactl.RunPromote(os.Stdout, karmadaConfig, opts, args)
+				gomega.Expect(err).Should(gomega.HaveOccurred())
```
Comment (XiShanYongYe-Chang) [inline test/e2e/karmadactl_test.go:83]: Error info can be varied, can you judge it more specifically? Alternatively, we may not test this error scenario.

File: test/e2e/karmadactl_test.go [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+				args := []string{"namespace", deploymentNamespace}
+				//init args: place namespace name to CommandPromoteOption.name
+				err := opts.Complete(args)
+				gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+
+				//use karmadactl to promote a namespace from member1
+				err = karmadactl.RunPromote(os.Stdout, karmadaConfig, opts, args)
+				gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+
+				framework.WaitNamespacePresentOnClusters(framework.ClusterNames(), deploymentNamespace)
```
Comment (XiShanYongYe-Chang) [inline test/e2e/karmadactl_test.go:101]: Here, we can simply check whether the namespace appears on the control plane.

File: test/e2e/karmadactl_test.go [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+
+		ginkgo.It("Test promoting clusterrole and clusterrolebindings ", func() {
+
+			// Step1, create cluster role and clusterrolebinding on member1
+			ginkgo.By(fmt.Sprintf("Creating clusterrole and clusterrolebinding in member: %s", member1), func() {
+				framework.CreateClusterRole(member1Client, clusterRole)
+				framework.CreateClusterRoleBinding(member1Client, clusterRoleBinding)
+			})
+
+			// Step2, create cluster role and clusterrolebinding on member1
```
Comment (XiShanYongYe-Chang) [inline test/e2e/karmadactl_test.go:182]: This comment seems mismatched.

**Review by XiShanYongYe-Chang — COMMENTED** [review]
Good job!

/cc @lonelyCZ Can you help review it.

File: test/e2e/karmadactl_test.go [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+	"github.com/onsi/gomega"
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/rand"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"github.com/karmada-io/karmada/pkg/karmadactl"
+	"github.com/karmada-io/karmada/pkg/karmadactl/options"
+	"github.com/karmada-io/karmada/test/e2e/framework"
+	"github.com/karmada-io/karmada/test/helper"
```
[the remark this answers is not in the document]
Comment (wuyingjun-lucky) [reply test/e2e/karmadactl_test.go:27]: thanks for the Guidance

File: test/e2e/karmadactl_test.go (hunk 2 above)
Comment (wuyingjun-lucky) [reply test/e2e/karmadactl_test.go:83]: K8S api server will return a not found error but karmadactl convert it to normal .  I can remove it here

**Comment by lonelyCZ** [comment]
Thanks, I will review it ASAP.

File: test/e2e/karmadactl_test.go [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+				gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+
+				// use karmadactl to promote a namespace from member1
+				err = karmadactl.RunPromote(os.Stdout, karmadaConfig, opts, args)
+				gomega.Expect(err).ShouldNot(gomega.HaveOccurred())
+
+				framework.WaitNamespacePresentOnClusterByClient(kubeClient, deploymentNamespace)
+			})
+
+			// Step 3,  promote deploy from cluster member1 to karmada
```
Comment (lonelyCZ) [inline test/e2e/karmadactl_test.go:82]:
```suggestion
			// Step 3,  promote deployment from cluster member1 to karmada
```

File: test/e2e/karmadactl_test.go [hunk 7]
```diff
@@ -0,0 +1,307 @@
...
+
+		ginkgo.AfterEach(func() {
+			framework.RemoveClusterRole(kubeClient, clusterRoleName)
+			framework.RemoveClusterPropagationPolicy(karmadaClient, clusterRoleName+"-propagation")
+
+			framework.RemoveClusterRoleBinding(kubeClient, clusterRoleBindingName)
+			framework.RemoveClusterPropagationPolicy(karmadaClient, clusterRoleBindingName+"-propagation")
+		})
+
+		ginkgo.It("Test promoting clusterrole and clusterrolebindings ", func() {
```
Comment (lonelyCZ) [inline test/e2e/karmadactl_test.go:151]:
```suggestion
		ginkgo.It("Test promoting clusterrole and clusterrolebindings", func() {
```

File: test/e2e/karmadactl_test.go [hunk 8]
```diff
@@ -0,0 +1,307 @@
...
+					return err1 == nil && err2 == nil
+				}, pollTimeout, pollInterval).Should(gomega.Equal(true))
+			})
+		})
+
+	})
+})
+
+var _ = framework.SerialDescribe("Karmadactl unjoin testing", ginkgo.Labels{NeedCreateCluster}, func() {
+	ginkgo.Context(" unjoining not ready cluster", func() {
```
Comment (lonelyCZ) [inline test/e2e/karmadactl_test.go:199]:
```suggestion
	ginkgo.Context("unjoining not ready cluster", func() {
```

File: test/e2e/karmadactl_test.go [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+
+		ginkgo.AfterEach(func() {
+			framework.RemoveDeployment(kubeClient, deploymentNamespace, deploymentName)
+			framework.RemovePropagationPolicy(karmadaClient, deploymentNamespace, deploymentName+"-propagation")
+			framework.RemoveNamespace(kubeClient, deploymentNamespace)
+		})
+
+		ginkgo.It("Test promoting a deployment from cluster member", func() {
+
+			// Step 1,  create ns and deploy on cluster member1
```
Comment (lonelyCZ) [inline test/e2e/karmadactl_test.go:57]:
```suggestion
			// Step 1,  create namespace and deployment on cluster member1
```

File: test/e2e/karmadactl_test.go [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+
+		ginkgo.It("Test promoting clusterrole and clusterrolebindings ", func() {
+
+			// Step1, create clusterrole and clusterrolebinding on member1
+			ginkgo.By(fmt.Sprintf("Creating clusterrole and clusterrolebinding in member: %s", member1), func() {
+				framework.CreateClusterRole(member1Client, clusterRole)
+				framework.CreateClusterRoleBinding(member1Client, clusterRoleBinding)
+			})
+
+			// Step2, promote cluster role and clusterrolebinding on member1
```
Comment (lonelyCZ) [inline test/e2e/karmadactl_test.go:159]:
```suggestion
			// Step2, promote clusterrole and clusterrolebinding from member1
```

File: test/e2e/karmadactl_test.go [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+)
+
+var _ = ginkgo.Describe("Karmadactl promote testing", func() {
+	var karmadaConfig karmadactl.KarmadaConfig
+	var member1 string
+	var member1Client kubernetes.Interface
+
+	ginkgo.BeforeEach(func() {
+		karmadaConfig = karmadactl.NewKarmadaConfig(clientcmd.NewDefaultPathOptions())
+		member1 = framework.ClusterNames()[0]
```
Comment (lonelyCZ) [inline test/e2e/karmadactl_test.go:35]: Perhaps, we should not define variate as specific cluster name, we also add some e2e test for PULL mode cluster like `member3`.

**Review by lonelyCZ — COMMENTED** [review]
Overall very good, a little modification suggestions

File: test/e2e/karmadactl_test.go (hunk 11 above)
Comment (wuyingjun-lucky) [reply test/e2e/karmadactl_test.go:35]: member1 = "member1" or directly use "member1"

Comment (lonelyCZ) [reply test/e2e/karmadactl_test.go:35]: I just looked it, directly using `member1 = "member1"` is ok. I think we also can provide some e2e tests for `member3` in the feature..

Comment (wuyingjun-lucky) [reply test/e2e/karmadactl_test.go:35]: > 
OK; OK

**Comment by wuyingjun-lucky** [comment]
@XiShanYongYe-Chang  The last comments is resolved

**Comment by XiShanYongYe-Chang** [comment]
@wuyingjun-lucky Ok. cc @lonelyCZ

**Comment by lonelyCZ** [comment]
/lgtm

suggest to mark this pr as `/kind feature`

**Review by XiShanYongYe-Chang — COMMENTED** [review]
/approve

**Pull request merged**